network/connecteventmanager.go

```diff
@@ -2,104 +2,213 @@ package network
 
 import (
 	"sync"
+	"time"
 
 	"github.com/libp2p/go-libp2p-core/peer"
 )
 
+type connectEvent int
+
+const (
+	connectEventConnected connectEvent = iota
+	connectEventDisconnected
+	connectEventMarkUnresponsive
+	connectEventMessage
+)
+
+type connEvt struct {
+	p       peer.ID
+	evtType connectEvent
+}
+
 type ConnectionListener interface {
 	PeerConnected(peer.ID)
 	PeerDisconnected(peer.ID)
 }
 
 type connectEventManager struct {
 	connListener ConnectionListener
-	lk           sync.RWMutex
-	conns        map[peer.ID]*connState
-}
-
-type connState struct {
-	refs       int
-	responsive bool
+	events       chan connEvt
+	done         chan struct{}
 }
 
 func newConnectEventManager(connListener ConnectionListener) *connectEventManager {
 	return &connectEventManager{
 		connListener: connListener,
-		conns:        make(map[peer.ID]*connState),
+		events:       make(chan connEvt, 256),
+		done:         make(chan struct{}),
 	}
 }
 
 func (c *connectEventManager) Connected(p peer.ID) {
-	c.lk.Lock()
-	defer c.lk.Unlock()
+	c.events <- connEvt{p, connectEventConnected}
+}
 
-	state, ok := c.conns[p]
-	if !ok {
-		state = &connState{responsive: true}
-		c.conns[p] = state
-	}
-	state.refs++
+func (c *connectEventManager) Disconnected(p peer.ID) {
+	c.events <- connEvt{p, connectEventDisconnected}
+}
 
-	if state.refs == 1 && state.responsive {
-		c.connListener.PeerConnected(p)
+func (c *connectEventManager) MarkUnresponsive(p peer.ID) {
+	c.events <- connEvt{p, connectEventMarkUnresponsive}
+}
+
+func (c *connectEventManager) OnMessage(p peer.ID) {
+	c.events <- connEvt{p, connectEventMessage}
+}
+
+func (c *connectEventManager) Run() {
+	gcTicker := time.NewTicker(30 * time.Second)
+	defer gcTicker.Stop()
+
+	peerEventsHandlers := make(map[peer.ID]*peerEventsHandler)
+	defer func() {
+		for _, h := range peerEventsHandlers {
+			h.shutdown()
+		}
+	}()
+
+	for {
+		select {
+		case evt := <-c.events:
+			handler, ok := peerEventsHandlers[evt.p]
+			if !ok {
+				if evt.evtType != connectEventConnected {
+					break
+				}
+				handler = newPeerEventsHandler(evt.p, c.connListener)
+				go handler.run()
+				peerEventsHandlers[evt.p] = handler
+			}
+			handler.add(evt.evtType)
+
+		case <-gcTicker.C:
```

Comment:
I'm not a fan of this approach. Can we have an un-buffered channel where each event handler writes to this unbuffered channel when it's ready to shutdown? E.g.:

```go
// In the handler:
// we're now ready to stop
select {
    case c.doneCh<-p:
    case evt <- pe.events:
    // nevermind, we're not stopping.
}

// in this goroutine:

select {
...
...
case <-p.done:
    ... remove it...
}
```

Reply:
The problem is that when the handler thinks it's ready to shut down, the main event loop may be in the middle of passing it a new event (eg new connection) that would cause it to not shut down.
So the only safe way I could think of is for the main loop to periodically ask the handler if it's done. That way the main loop can ask the handler if it's done when the main loop is not actively giving events to the handler.

Thinking a little more about it though, we could implement it like this:
1. when the handler's ref count reaches zero it signals the main loop
2. when the main loop finishes processing incoming events it
  - sees the signal
  - asks the handler if it's still ready to shut down
    - if so, deletes and shuts down the handler
    - if not does nothing, and the handler waits for the next disconnect event

Reply:
Take a look at my suggested code again. I believe it addresses that problem by using a synchronous channel. If the event loop is trying to pass an event to the handler, the handler will take the event and stop trying to shutdown.

Reply:
Ah yes you're right I think that should work. We just need to make sure that in the case where we get for example `Disconnect` which is interrupted by a subsequent `Message` event, that we still try to shut down after the `Message` event.

Reply:
I think that the suggested code above is actually not safe in the following scenario:
- Main thread:
  1. Receive `Disconnect` for Peer A
  2. Receive `Connect` for Peer A
  3. No more events: select on more events / handler done
- Handler for Peer A:
  1. Channel receives `Disconnect` event
  2. Channel receives `Connect` event
  3. Handler thread pops `Disconnect` event from channel
    Handler thread reaches the select:
```
select {
    case c.doneCh<-p:
    case evt <- pe.events:
    // nevermind, we're not stopping.
}
```
  At this point it's possible that the first case could succeed, and the main thread cleans up the handler even though the `Connect` event has been processed.

Reply:
Note: It _would_ work if the handler's channel is unbuffered, but I don't think we want that because then the main thread is blocked while the handler threads are calling `Connect()` on Bitswap.

Reply:
Ah, I see. Yeah, either:

1. The channel would have to be unbuffered.
2. The outer loop would have to check the channel itself.

Reply:
The idea I'm proposing is 2. The handler keeps track of when things are added to the channel / processed, and when
- the event is a `Disconnect`
- there's nothing in the channel
it signals the main thread

The main thread checks if the channel is still in the same state, and if so, cleans it up. If not it waits for another signal.